katz/src/main/kotlin/katz/Reader.kt

```diff
@@ -0,0 +1,38 @@
+package kats
+
+/**
+ * Basic implementation of the Reader monad. Provides an "implicit" context (configuration) for
+ * function execution. Intended to provide Dependency Injection.
+ */
+class Reader<C : Any, out A : Any>(val run: (C) -> A) {
+
+    inline fun <B : Any> map(crossinline fa: (A) -> B): Reader<C, B> = Reader { c -> fa(run(c)) }
+
+    inline fun <B : Any> flatMap(crossinline fa: (A) -> Reader<C, B>): Reader<C, B> = Reader { c ->
```

Comment:
What's the benefit on declaring `C` as contravariant ? I have just added `zip` combinator, where `C` is used again as a covariant type (like in flatMap), so not really sure about whether it still makes sense to make it contravariant.

Can you share your opinion on this @raulraja ?

Oh, and your extension method doesn't seem to work properly with the usage I do of `flatMap` in tests. Your approach seem to lack something @JMPergar.